src/ol/style/fillstyle.js

```diff
@@ -57,8 +58,14 @@ ol.style.Fill.prototype.setColor = function(color) {
  */
 ol.style.Fill.prototype.getChecksum = function() {
   if (this.checksum_ === undefined) {
-    this.checksum_ = 'f' + (this.color_ ?
-        ol.color.asString(this.color_) : '-');
+    if (this.color_ instanceof CanvasPattern) {
+      this.checksum_ = 'fCanvasPattern';
```

Comment:
You should be able to use `goog.getUid(this.color_)` for the canvas pattern and gradient.